fabtests/functional/bw.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2019 Intel Corporation.  All rights reserved.
+ *
+ * This software is available to you under the BSD license
+ * below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+
+#include <shared.h>
+
+static ssize_t post_one_tx(struct ft_context *msg)
+{
+	if (ft_check_opts(FT_OPT_VERIFY_DATA | FT_OPT_ACTIVE))
+		ft_fill_buf(msg->buf + ft_tx_prefix_size(),
+			    opts.transfer_size);
+
+	return ft_post_tx_buf(ep, remote_fi_addr, opts.transfer_size,
+			      NO_CQ_DATA, &msg->context, msg->buf,
+			      msg->desc, 0);
+}
+
+static ssize_t wait_check_rx_bufs(void)
+{
+	ssize_t ret;
+	int i;
+	struct ft_context *context;
+
+	ret = ft_get_rx_comp(rx_seq);
+	if (ret)
+		return ret;
+
+	if (!ft_check_opts(FT_OPT_VERIFY_DATA | FT_OPT_ACTIVE))
+		return 0;
+
+	for (i = 0; i < opts.window_size; i++) {
+		context = ft_get_ctx_index(&rx_ctx_arr, i, rx_mr_size);
+		if (!context)
+			return -FI_EINVAL;
+
+		ret = ft_check_buf((char *) context->buf +
+				   ft_rx_prefix_size(), opts.transfer_size);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+static int run_loop(void)
+{
+	int ret, i;
+	struct ft_context *context;
+
+	ft_start();
+	if (opts.dst_addr) {
+		for (i = 0; i < opts.window_size; i++) {
+			context = ft_get_ctx_index(&tx_ctx_arr, i, tx_mr_size);
+			if (!context)
+				return -FI_EINVAL;
+			ret = post_one_tx(context);
+			if (ret)
+				return ret;
+		}
+		ret = ft_sync_manual();
+		if (ret)
+			return ret;
+
+		ret = ft_get_tx_comp(tx_seq);
+		if (ret)
+			return ret;
+	} else {
+		ret = ft_sync_manual();
```

Comment:
And I would move this down until after we've posted all messages and gotten their completions.